internal/overlord/checkstate/manager.go

```diff
@@ -0,0 +1,216 @@
+// Package checkstate is the manager and "checkers" for custom health checks.
+package checkstate
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+)
+
+// CheckManager starts and manages the health checks.
+type CheckManager struct {
+	mutex           sync.Mutex
+	checks          map[string]*checkData
+	failureHandlers []FailureFunc
+}
+
+type FailureFunc func(name string)
+
+// NewManager creates a new check manager.
+func NewManager() *CheckManager {
+	return &CheckManager{}
+}
+
+// AddFailureHandler adds f to the list of "failure handlers", functions that
+// are called whenever a check hits its failure threshold.
+func (m *CheckManager) AddFailureHandler(f FailureFunc) {
+	m.failureHandlers = append(m.failureHandlers, f)
+}
+
+// Configure handles updates to the plan (server configuration),
+// stopping the previous checks and starting the new ones as required.
+func (m *CheckManager) Configure(p *plan.Plan) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	logger.Debugf("Configuring check manager (stopping %d, starting %d)",
+		len(m.checks), len(p.Checks))
+
+	// First stop existing checks.
+	for _, check := range m.checks {
+		check.stop()
+	}
+
+	// Then configure and start new checks.
+	checks := make(map[string]*checkData, len(p.Checks))
+	for name, config := range p.Checks {
+		check := &checkData{
+			config:  config,
+			checker: newChecker(config),
+			done:    make(chan struct{}),
+			action:  m.callFailureHandlers,
+		}
+		checks[name] = check
+		go check.loop()
+	}
+	m.checks = checks
+}
+
+func (m *CheckManager) callFailureHandlers(name string) {
+	for _, f := range m.failureHandlers {
+		f(name)
+	}
+}
+
+// Checks returns the list of currently-configured checks and their status,
+// ordered by name.
+//
+// If level is not UnsetLevel, the list of checks is filtered to only include
+// checks with the given level. If names is non-empty, the list of checks is
+// filtered to only include the named checks.
+func (m *CheckManager) Checks(level plan.CheckLevel, names []string) ([]*CheckInfo, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	var infos []*CheckInfo
+	for _, check := range m.checks {
+		if level != plan.UnsetLevel && level != check.config.Level {
+			continue
+		}
+		if len(names) > 0 && !containsString(names, check.config.Name) {
```

Comment:
looks like containsString is used once, probably just inline it here

Reply:
I think it's a bit nicer as a stand-alone function, as it allows a simpler early return rather than something like `break containsLoop`. However, I found that we already have a `strutil.ListContains` function, so I changed it to use that. Likely in Go 1.19 we'll be able to write `slices.Contains` (in Go 1.18 `slices` will be an experimental generics-based package).